src/Dialect/ONNX/ONNXOpsHelper.cpp

```diff
@@ -357,11 +358,13 @@ bool AreTheSameConstantOpDenseAttr(
 /// Test if 'val' has shape and rank or not.
 bool hasShapeAndRank(Value val) {
   Type valType = val.getType();
-  if (!valType.isa<SeqType>())
-    return valType.isa<ShapedType>() && valType.cast<ShapedType>().hasRank();
-
-  // Check the element type for a Sequence.
-  return valType.cast<SeqType>().getElementType().hasRank();
+  ShapedType shapedType;
+  if (auto seqType = valType.dyn_cast<SeqType>()) {
+    shapedType = seqType.getElementType().dyn_cast<ShapedType>();
```

Comment:
How about Seq(Map(T1, T2))? It's unclear to me what's the usage of the function hasShapeAndRank.

Reply:
if valType is Seq(Map(T1, T2)) then seqType.getElementType() is Map(T1, T2) and dyn_cast<ShapedType>() returns nullptr and hasShapeAndRank() returns false, which I think follows the spirit of the previous implementation and use of hasShapeAndRank

yeah, hasShapeAndRank is a bit strange and the way it's used in many verify() and inferShapes() implementations in ONNXOps.cpp could probably be done more cleanly using functions from src/Support/TypeUtilities